src/libraries/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -114,6 +117,7 @@ public static void RegisterFactory(string providerInvariantName, string factoryT
             _registeredFactories[providerInvariantName] = new ProviderRegistration(factoryTypeAssemblyQualifiedName, null);
         }
 
+        [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
```

Comment:
I wasn't sure if linker had enough info here to preserve the type and members - once the type is instantiated we do not know what methods will be called. I'm not really sure, would `DynamicallyAccessedMemberTypes.All` on the type help here?

Reply:
Ok, so I looked through this code a bit more, and I think we should be re-doing how we are annotating this class.

What I can tell this code is doing:

1. There are "Register" and "Get" methods. "Register" puts info about the provider into a dictionary. "Get" gets the info out of the dictionary, and returns the Provider.
2. There are 3 pieces of information being stored:
    1. A  "providerInvariantName" - cache key
    2. The Provider Type's assembly qualified name
    3. An optional Provider instance
3. In order to "Get" a provider, we do the following:
    1. Check if the instance is already on the registration, if it is return it
    2. If not, we get the Type from the assembly qualified name
    3. Use Reflection to get the `Instance` field off the Type

What I think should be done instead is:

1. We annotate the `ProviderRegistration.FactoryTypeAssemblyQualifiedName` property and any string parameters that go into this property as `[DAM(PublicFields)]`.
2. Remove the `UnconditionalSuppressMessage` on `private static DbProviderFactory? GetFactory(string providerInvariantName, bool throwOnError)`.
    1. This should no longer be necessary because the Type and `string`s being used in this method should now all be annotated correctly.
    2. We can also remove the `M:System.Data.Common.DbProviderFactories.GetFactoryInstance(System.Type)` entry in the ILLink.Suppressions.xml file

This will now make this class 99% safe. The correct types will be annotated on the way "in" during Register. The correct things will be preserved. On the way "out" during Get, even when we need to use Reflection, the registration information is annotated, and the trimmer will no longer warn.

The only place that is unsafe is the `DataRow` overload, because there is no way to annotate the values of a column in that row as `[DAM]`. So that will remain `RequiresUnreferencedCode`.

Reply:
This is neat, I'm still getting one more warning with this and not sure how to fix it:

```csharp
        public static void RegisterFactory(string providerInvariantName, DbProviderFactory factory)
        {
            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
            ADP.CheckArgumentNull(factory, nameof(factory));

            string factoryTypeAssemblyQualifiedName = factory.GetType().AssemblyQualifiedName!; // <-------
            _registeredFactories[providerInvariantName] = new ProviderRegistration(factoryTypeAssemblyQualifiedName, factory);
        }
```

```
d:\src\runtime-2\src\libraries\System.Data.Common\src\System\Data\Common\DbProviderFactories.cs(130,13): Trim analysis warning IL2072: System.Data.Common.DbProviderFactories.RegisterFactory(String,DbProviderFactory): 'factoryTypeAssemblyQualifiedName' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields' in call to 'System.Data.Common.DbProviderFactories.ProviderRegistration.ProviderRegistration(String,DbProviderFactory)'. The return value of method 'System.Type System.Object::GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [d:\src\runtime-2\src\libraries\src.proj]
```

trimmer doesn't seem to be able to connect that this instance is related to this type. Should this be just suppressed and perhaps DynamicDependency added?

Reply:
For this to work you will need to add [DAM(PublicFields)] onto the DbProviderFactory type itself (so that it and all its derived types preserve public fields). Then I "think" this should work.

Reply:
The other routes I was thinking here is to either:

1. Suppress the warning, since we already have the instance we need - it was provided in the argument. The danger here is that we are still annotating `FactoryTypeAssemblyQualifiedName` as `[DAM(PublicField)]`, so usages of that property could still try getting the type and expect to read the properties. The only place in existing code that reads this property even if there is an instance registered is in `public static DataTable GetFactoryClasses()`, which the string would get stuck in a `DataTable`, which would be unannotated anyway. But if we ever made changes to this code, we would need to be aware of this suppression, so it probably isn't a good approach.

2. When we have an instance, like we do here, don't fill out the `FactoryTypeAssemblyQualifiedName` property and leave it `null`. However, this could break other things, like the `public static DataTable GetFactoryClasses()` method. So we'd have to change that method to detect when `FactoryTypeAssemblyQualifiedName` is null, and grab the Type from the FactoryInstance.

If @vitek-karas's suggestion works, that would probably be the best/easiest.

Reply:
@vitek-karas's suggestion works! Thank you!